### PR TITLE
[clangd] Fix is spelled in source bug

### PR DESCRIPTION
**Pull request by schenka0**
This fixes the issue reported in #76667 and adds an initial unit test for isSpelledInSource().

Note that in that issue there was still some underlying corrupted AST, but this at least makes isSpelledInSource() robust to it.

### CODE REVIEW

**Comment by schenka0** [comment]
PR: @sam-mccall

File: clang-tools-extra/clangd/SourceCode.cpp [hunk 1]
```diff
@@ -232,7 +232,12 @@ bool isSpelledInSource(SourceLocation Loc, const SourceManager &SM) {
   if (Loc.isFileID())
     return true;
   auto Spelling = SM.getDecomposedSpellingLoc(Loc);
-  StringRef SpellingFile = SM.getSLocEntry(Spelling.first).getFile().getName();
+  bool InvalidSLocEntry = false;
+  const auto SLocEntry = SM.getSLocEntry(Spelling.first, &InvalidSLocEntry);
+  if (InvalidSLocEntry) {
+    return false;
+  }
```
Comment (zyn0217) [inline clang-tools-extra/clangd/SourceCode.cpp:239]:
```suggestion
  if (InvalidSLocEntry)
    return false;
```
We don't usually add braces to one-line-if statements.

See https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements.

File: clang-tools-extra/clangd/SourceCode.cpp [hunk 2]
```diff
@@ -232,7 +232,12 @@ bool isSpelledInSource(SourceLocation Loc, const SourceManager &SM) {
...
+  const StringRef SpellingFile = SLocEntry.getFile().getName();
```
Comment (zyn0217) [inline clang-tools-extra/clangd/SourceCode.cpp:240]: The const qualifier looks unnecessary to me. Could you please leave it out?

**Review by zyn0217 — COMMENTED** [review]
Thanks for your contribution! A couple of nit-picks, and I’d leave the approval to other folks.

File: clang-tools-extra/clangd/unittests/SourceCodeTests.cpp [hunk 3]
```diff
@@ -813,6 +813,21 @@ TEST(SourceCodeTests, isKeywords) {
   EXPECT_FALSE(isKeyword("override", LangOpts));
 }
 
+TEST(SourceCodeTests, isSpelledInSource) {
+  Annotations Test(R"cpp(
+        int abc = 1;
+    )cpp");
+
+  ParsedAST AST = TestTU::withCode(Test.code()).build();
+  llvm::errs() << Test.code();
```
[the remark this answers is not in the document]
Comment (schenka0) [reply clang-tools-extra/clangd/unittests/SourceCodeTests.cpp]: Thanks for the review. Not familiar with this codebase/standards so was just copying surrounding code. I've simplified the test code.

**Comment by schenka0** [comment]
I struggled to rebase my changes around some merges and accidently included additional changes at one point which triggered the llvmbot to add extra tags. Sorry!

File: clang-tools-extra/clangd/unittests/SourceCodeTests.cpp [hunk 4]
```diff
@@ -813,6 +813,18 @@ TEST(SourceCodeTests, isKeywords) {
...
 }
 
+TEST(SourceCodeTests, isSpelledInSource) {
+  Annotations Test("");
+  ParsedAST AST = TestTU::withCode(Test.code()).build();
+  const SourceManager &SM = AST.getSourceManager();
+
+  EXPECT_TRUE(
+      isSpelledInSource(SM.getLocForStartOfFile(SM.getMainFileID()), SM));
+  EXPECT_TRUE(isSpelledInSource(SourceLocation(), SM));
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/unittests/SourceCodeTests.cpp:823]: I think it's worth adding a couple of comments here.

First:

```c++
// Check that isSpelledInSource() handles various invalid source locations gracefully.
```

But also, a slightly more subtle point:

```c++
// Returning true for SourceLocation() is a behavior that falls out of the current
// implementation, which has an early exit for isFileID().
// FIXME: Should it return false on SourceLocation()? Does it matter?
```

**Review by HighCommander4 — APPROVED** [review]
Thanks, looks good with the added comments.

**Comment by schenka0** [comment]
@HighCommander4 Thanks for the review. I've updated the comments, but don't have merge access, could you merge this? Thanks

**Review by HighCommander4 — APPROVED** [review]
Thanks!

**Comment by HighCommander4** [comment]
> @HighCommander4 Thanks for the review. I've updated the comments, but don't have merge access, could you merge this? Thanks

Done

**Comment by zyn0217** [comment]
(A drive-by comment: I stumbled on the commit message of this PR and found it surprising that I'm the co-author of it! I'm not entirely clear on GitHub's schemas, but IMO shouldn't the co-author be the one who clicks the "merge" button?)

**Comment by HighCommander4** [comment]
> (A drive-by comment: I stumbled on the commit message of this PR and found it surprising that I'm the co-author of it! I'm not entirely clear on GitHub's schemas, but IMO shouldn't the co-author be the one who clicks the "merge" button?)

The reason is that one of the four commits that were squashed had you as a co-author:

image

(I'm not sure how that happened in the first place.)

**Pull request merged**